presto-main/src/main/java/com/facebook/presto/cost/RowNumberStatsRule.java

```diff
@@ -0,0 +1,90 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.Patterns;
+import com.facebook.presto.sql.planner.plan.RowNumberNode;
+
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.min;
+
+public class RowNumberStatsRule
+        extends SimpleStatsRule<RowNumberNode>
+{
+    private static final Pattern<RowNumberNode> PATTERN = Patterns.rowNumber();
+
+    public RowNumberStatsRule(StatsNormalizer normalizer)
+    {
+        super(normalizer);
+    }
+
+    @Override
+    public Pattern<RowNumberNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> doCalculate(RowNumberNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        if (sourceStats.isOutputRowCountUnknown()) {
+            return Optional.empty();
+        }
+
+        double partitionCount = 1;
+        for (Symbol groupBySymbol : node.getPartitionBy()) {
+            SymbolStatsEstimate symbolStatistics = sourceStats.getSymbolStatistics(groupBySymbol);
+            int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
+            // assuming no correlation between grouping keys
+            partitionCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
+        }
+        partitionCount = min(sourceStats.getOutputRowCount(), partitionCount);
+
+        if (isNaN(partitionCount)) {
+            return Optional.empty();
+        }
+
+        double sourceRowsCount = sourceStats.getOutputRowCount();
+        double rowsPerPartition = sourceRowsCount / partitionCount;
+        if (node.getMaxRowCountPerPartition().isPresent()) {
+            rowsPerPartition = min(rowsPerPartition, node.getMaxRowCountPerPartition().get());
+        }
+
+        double outputRowsCount = sourceStats.getOutputRowCount();
+        if (node.getMaxRowCountPerPartition().isPresent()) {
+            outputRowsCount = partitionCount * rowsPerPartition;
+        }
+
+        return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRowNumberSymbol(), SymbolStatsEstimate.builder()
+                        // setting only well known low value
+                        // it is hard to estimate high value correctly
+                        .setLowValue(1)
+                        .setDistinctValuesCount(rowsPerPartition)
```

Comment:
for row count symbol, `NDV = high - low + 1` (always). So, "high value is known" ⇔ "NDV is known".

Reply:
The `NDV` is an expected value. When total number of rows is N, and the number of expected partitions is M, the expected number of distinct values is N / M. However, as you pointed out earlier we shouldn't set High value, as it has stricter requirements.

Reply:
I don't understand this. What are the stricter requirements?

> The NDV is an expected value. 

Right, but you're not estimating an expected NDV.
Each partition will generate values `[1, 2, 3, ... rowsInThisPartition]`

If all partitions are equal in size (no skew), then each partition will have `rowsInThisPartition == rowsPerPartition` and then `rowNumberSymbol` will have `rowsPerPartition` unique values (NDV = `rowsPerPartition`). If there is any skew, some partitions will be smaller and some will be larger. The largest partition being the one which defines NDV.

Thus, `rowsPerPartition` is a lower bound (achievable / possible) of NDV, so hardly the expected value.
If we assume no skew (which we often do), this may stay as is, but deserves a code comment like the following (if this is what you mean here):

```
// Set NDV assuming no skew.
// Note: if we assume no skew, we could also estimate High value (as rowsPerPartition), but 
// underestimation of High value may have more severe consequences than underestimation of NDV.
.setDistinctValuesCount(rowsPerPartition)
```

Reply:
> so hardly the expected value

Yes. You are right. This is the value assuming no skew. Let me add a comment.